File: src/core/wasm/SpatialIndex.ts
/**
 * ============================================================================
 * SPATIAL INDEX (DYNAMIC 2D QUADTREE FOR NODE & WIRE HIT-TESTING)
 * ============================================================================
 * Implements high-performance spatial indexing for visual graph nodes & wires:
 * - Incremental updates avoiding full rebuild on each frame
 * - O(log N) hit testing for clicks, marquee drag select, and viewport culling
 * - Fast point queries and AABB bounding range queries
 * ============================================================================
 */

export interface AABB {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
}

export type SpatialItemType = "node" | "pin" | "wire" | "comment";

export interface SpatialItem {
  id: string;
  type: SpatialItemType;
  bounds: AABB;
  zIndex: number;
}

export class QuadtreeNode {
  public static readonly MAX_ITEMS = 8;
  public static readonly MAX_DEPTH = 8;

  public bounds: AABB;
  public depth: number;
  public isSubdivided: boolean = false;
  public items: SpatialItem[] = [];
  public children: [QuadtreeNode, QuadtreeNode, QuadtreeNode, QuadtreeNode] | null = null;

  constructor(bounds: AABB, depth: number = 0) {
    this.bounds = bounds;
    this.depth = depth;
  }

  public static containsPoint(bounds: AABB, x: number, y: number): boolean {
    return x >= bounds.minX && x <= bounds.maxX && y >= bounds.minY && y <= bounds.maxY;
  }

  public static intersects(a: AABB, b: AABB): boolean {
    return !(a.minX > b.maxX || a.maxX < b.minX || a.minY > b.maxY || a.maxY < b.minY);
  }

  public subdivide(): void {
    const xMid = (this.bounds.minX + this.bounds.maxX) * 0.5;
    const yMid = (this.bounds.minY + this.bounds.maxY) * 0.5;
    const d = this.depth + 1;

    this.children = [
      new QuadtreeNode({ minX: this.bounds.minX, minY: this.bounds.minY, maxX: xMid, maxY: yMid }, d), // NW
      new QuadtreeNode({ minX: xMid, minY: this.bounds.minY, maxX: this.bounds.maxX, maxY: yMid }, d), // NE
      new QuadtreeNode({ minX: this.bounds.minX, minY: yMid, maxX: xMid, maxY: this.bounds.maxY }, d), // SW
      new QuadtreeNode({ minX: xMid, minY: yMid, maxX: this.bounds.maxX, maxY: this.bounds.maxY }, d), // SE
    ];
    this.isSubdivided = true;

    const remaining: SpatialItem[] = [];
    for (const item of this.items) {
      let placed = false;
      for (const child of this.children) {
        if (
          QuadtreeNode.containsPoint(child.bounds, item.bounds.minX, item.bounds.minY) &&
          QuadtreeNode.containsPoint(child.bounds, item.bounds.maxX, item.bounds.maxY)
        ) {
          child.insert(item);
          placed = true;
          break;
        }
      }
      if (!placed) {
        remaining.push(item);
      }
    }
    this.items = remaining;
  }

  public insert(item: SpatialItem): boolean {
    if (!QuadtreeNode.intersects(this.bounds, item.bounds)) {
      return false;
    }

    if (this.isSubdivided && this.children) {
      for (const child of this.children) {
        if (
          QuadtreeNode.containsPoint(child.bounds, item.bounds.minX, item.bounds.minY) &&
          QuadtreeNode.containsPoint(child.bounds, item.bounds.maxX, item.bounds.maxY)
        ) {
          return child.insert(item);
        }
      }
      this.items.push(item);
      return true;
    }

    this.items.push(item);

    if (this.items.length > QuadtreeNode.MAX_ITEMS && this.depth < QuadtreeNode.MAX_DEPTH) {
      this.subdivide();
    }

    return true;
  }

  public remove(id: string): boolean {
    const idx = this.items.findIndex((item) => item.id === id);
    let found = false;
    if (idx !== -1) {
      this.items.splice(idx, 1);
      found = true;
    }

    if (this.isSubdivided && this.children) {
      for (const child of this.children) {
        if (child.remove(id)) {
          found = true;
        }
      }
    }

    return found;
  }

  public queryRange(range: AABB, results: SpatialItem[]): void {
    if (!QuadtreeNode.intersects(this.bounds, range)) {
      return;
    }

    for (const item of this.items) {
      if (QuadtreeNode.intersects(item.bounds, range)) {
        results.push(item);
      }
    }

    if (this.isSubdivided && this.children) {
      for (const child of this.children) {
        child.queryRange(range, results);
      }
    }
  }

  public queryPoint(x: number, y: number, results: SpatialItem[]): void {
    if (!QuadtreeNode.containsPoint(this.bounds, x, y)) {
      return;
    }

    for (const item of this.items) {
      if (QuadtreeNode.containsPoint(item.bounds, x, y)) {
        results.push(item);
      }
    }

    if (this.isSubdivided && this.children) {
      for (const child of this.children) {
        child.queryPoint(x, y, results);
      }
    }
  }

  public clear(): void {
    this.items = [];
    this.isSubdivided = false;
    this.children = null;
  }

  public totalItemCount(): number {
    let count = this.items.length;
    if (this.isSubdivided && this.children) {
      for (const child of this.children) {
        count += child.totalItemCount();
      }
    }
    return count;
  }
}

export class SpatialIndex {
  private worldBounds: AABB;
  private root: QuadtreeNode;
  private registry: Map<string, SpatialItem> = new Map();

  constructor(worldBounds: AABB) {
    this.worldBounds = worldBounds;
    this.root = new QuadtreeNode(worldBounds, 0);
  }

  public insert(id: string, type: SpatialItemType, bounds: AABB, zIndex: number = 0): void {
    if (this.registry.has(id)) {
      this.root.remove(id);
    }
    const item: SpatialItem = { id, type, bounds, zIndex };
    this.registry.set(id, item);
    this.root.insert(item);
  }

  public update(id: string, newBounds: AABB): void {
    const existing = this.registry.get(id);
    if (existing) {
      existing.bounds = newBounds;
      this.root.remove(id);
      this.root.insert(existing);
    }
  }

  public remove(id: string): boolean {
    if (this.registry.has(id)) {
      this.registry.delete(id);
      return this.root.remove(id);
    }
    return false;
  }

  public queryRange(range: AABB): SpatialItem[] {
    const results: SpatialItem[] = [];
    this.root.queryRange(range, results);
    return results;
  }

  public queryPoint(x: number, y: number): SpatialItem[] {
    const results: SpatialItem[] = [];
    this.root.queryPoint(x, y, results);
    // Sort descending by z-index (topmost first)
    results.sort((a, b) => b.zIndex - a.zIndex);
    return results;
  }

  public hitTestTopmost(x: number, y: number): SpatialItem | null {
    const items = this.queryPoint(x, y);
    return items.length > 0 ? items[0] : null;
  }

  public clear(): void {
    this.registry.clear();
    this.root.clear();
  }

  public size(): number {
    return this.registry.size;
  }
}

File: DOCS/Initial/decisions/0001-wasm-archive/src/SpatialIndex.cpp
#include "SpatialIndex.hpp"

namespace WebAppEngine {

// ============================================================================
// QuadtreeNode Implementation
// ============================================================================

QuadtreeNode::QuadtreeNode(const AABB& bounds, int depth)
    : m_bounds(bounds), m_depth(depth) {}

void QuadtreeNode::subdivide() {
    float xMid = (m_bounds.minX + m_bounds.maxX) * 0.5f;
    float yMid = (m_bounds.minY + m_bounds.maxY) * 0.5f;

    // NW: [minX, minY] -> [xMid, yMid]
    m_children[0] = std::make_unique<QuadtreeNode>(AABB(m_bounds.minX, m_bounds.minY, xMid, yMid), m_depth + 1);
    // NE: [xMid, minY] -> [maxX, yMid]
    m_children[1] = std::make_unique<QuadtreeNode>(AABB(xMid, m_bounds.minY, m_bounds.maxX, yMid), m_depth + 1);
    // SW: [minX, yMid] -> [xMid, maxY]
    m_children[2] = std::make_unique<QuadtreeNode>(AABB(m_bounds.minX, yMid, xMid, m_bounds.maxY), m_depth + 1);
    // SE: [xMid, yMid] -> [maxX, maxY]
    m_children[3] = std::make_unique<QuadtreeNode>(AABB(xMid, yMid, m_bounds.maxX, m_bounds.maxY), m_depth + 1);

    m_isSubdivided = true;

    // Re-distribute existing items into children if they fit completely
    std::vector<SpatialItem> remainingItems;
    for (const auto& item : m_items) {
        bool placed = false;
        for (int i = 0; i < 4; ++i) {
            if (m_children[i]->m_bounds.intersects(item.bounds)) {
                // If it fits within a single child, insert it there
                if (m_children[i]->m_bounds.contains(item.bounds.minX, item.bounds.minY) &&
                    m_children[i]->m_bounds.contains(item.bounds.maxX, item.bounds.maxY)) {
                    m_children[i]->insert(item);
                    placed = true;
                    break;
                }
            }
        }
        if (!placed) {
            remainingItems.push_back(item);
        }
    }
    m_items = std::move(remainingItems);
}

bool QuadtreeNode::insert(const SpatialItem& item) {
    if (!m_bounds.intersects(item.bounds)) {
        return false;
    }

    if (m_isSubdivided) {
        for (int i = 0; i < 4; ++i) {
            if (m_children[i]->m_bounds.contains(item.bounds.minX, item.bounds.minY) &&
                m_children[i]->m_bounds.contains(item.bounds.maxX, item.bounds.maxY)) {
                return m_children[i]->insert(item);
            }
        }
        // Item straddles boundaries, store in this parent node
        m_items.push_back(item);
        return true;
    }

    m_items.push_back(item);

    if (static_cast<int>(m_items.size()) > MAX_ITEMS && m_depth < MAX_DEPTH) {
        subdivide();
    }

    return true;
}

bool QuadtreeNode::remove(const std::string& id) {
    auto it = std::remove_if(m_items.begin(), m_items.end(),
        [&id](const SpatialItem& item) { return item.id == id; });

    bool found = (it != m_items.end());
    if (found) {
        m_items.erase(it, m_items.end());
    }

    if (m_isSubdivided) {
        for (int i = 0; i < 4; ++i) {
            if (m_children[i]->remove(id)) {
                found = true;
            }
        }
    }

    return found;
}

void QuadtreeNode::queryRange(const AABB& range, std::vector<SpatialItem>& results) const {
    if (!m_bounds.intersects(range)) {
        return;
    }

    for (const auto& item : m_items) {
        if (item.bounds.intersects(range)) {
            results.push_back(item);
        }
    }

    if (m_isSubdivided) {
        for (int i = 0; i < 4; ++i) {
            m_children[i]->queryRange(range, results);
        }
    }
}

void QuadtreeNode::queryPoint(float x, float y, std::vector<SpatialItem>& results) const {
    if (!m_bounds.contains(x, y)) {
        return;
    }

    for (const auto& item : m_items) {
        if (item.bounds.contains(x, y)) {
            results.push_back(item);
        }
    }

    if (m_isSubdivided) {
        for (int i = 0; i < 4; ++i) {
            if (m_children[i]->m_bounds.contains(x, y)) {
                m_children[i]->queryPoint(x, y, results);
            }
        }
    }
}

void QuadtreeNode::clear() {
    m_items.clear();
    m_isSubdivided = false;
    for (int i = 0; i < 4; ++i) {
        m_children[i].reset();
    }
}

size_t QuadtreeNode::totalItemCount() const {
    size_t count = m_items.size();
    if (m_isSubdivided) {
        for (int i = 0; i < 4; ++i) {
            count += m_children[i]->totalItemCount();
        }
    }
    return count;
}

// ============================================================================
// SpatialIndex Implementation
// ============================================================================

SpatialIndex::SpatialIndex(const AABB& worldBounds)
    : m_worldBounds(worldBounds), m_root(std::make_unique<QuadtreeNode>(worldBounds, 0)) {}

void SpatialIndex::insert(const std::string& id, SpatialItemType type, const AABB& bounds, int zIndex) {
    // If already exists, remove first
    auto it = m_itemRegistry.find(id);
    if (it != m_itemRegistry.end()) {
        m_root->remove(id);
    }

    SpatialItem item{id, type, bounds, zIndex};
    m_itemRegistry[id] = item;
    m_root->insert(item);
}

void SpatialIndex::update(const std::string& id, const AABB& newBounds) {
    auto it = m_itemRegistry.find(id);
    if (it != m_itemRegistry.end()) {
        it->second.bounds = newBounds;
        m_root->remove(id);
        m_root->insert(it->second);
    }
}

bool SpatialIndex::remove(const std::string& id) {
    auto it = m_itemRegistry.find(id);
    if (it != m_itemRegistry.end()) {
        m_itemRegistry.erase(it);
        return m_root->remove(id);
    }
    return false;
}

std::vector<SpatialItem> SpatialIndex::queryRange(const AABB& range) const {
    std::vector<SpatialItem> results;
    m_root->queryRange(range, results);
    return results;
}

std::vector<SpatialItem> SpatialIndex::queryPoint(float x, float y) const {
    std::vector<SpatialItem> results;
    m_root->queryPoint(x, y, results);

    // Sort descending by z-index (topmost first)
    std::sort(results.begin(), results.end(), [](const SpatialItem& a, const SpatialItem& b) {
        return a.zIndex > b.zIndex;
    });

    return results;
}

bool SpatialIndex::hitTestTopmost(float x, float y, SpatialItem& outItem) const {
    auto items = queryPoint(x, y);
    if (items.empty()) {
        return false;
    }
    outItem = items.front();
    return true;
}

void SpatialIndex::clear() {
    m_itemRegistry.clear();
    m_root->clear();
}

} // namespace WebAppEngine
